src/base/QXmppHttpUploadIq.cpp

```diff
@@ -0,0 +1,262 @@
+    writer->writeAttribute("size", QString::number(d->size));
+    // content-type is optional
+    if (!d->contentType.isDefault() && d->contentType.isValid())
+        writer->writeAttribute("content-type", d->contentType.name());
+    writer->writeEndElement();
+}
+/// \endcond
+
+class QXmppHttpUploadSlotIqPrivate
+{
+public:
+    QUrl putUrl;
+    QUrl getUrl;
+    QMap<QString, QString> headerFields;
+};
+
+QXmppHttpUploadSlotIq::QXmppHttpUploadSlotIq()
+    : d(new QXmppHttpUploadSlotIqPrivate())
+{
+}
+
+QXmppHttpUploadSlotIq::~QXmppHttpUploadSlotIq()
+{
+    delete d;
+}
+
+/// Returns the URL for uploading via. HTTP PUT.
+
+QUrl QXmppHttpUploadSlotIq::putUrl() const
+{
+    return d->putUrl;
+}
+
+/// Sets the URL the client should use for uploading.
+
+void QXmppHttpUploadSlotIq::setPutUrl(const QUrl &putUrl)
+{
+    d->putUrl = putUrl;
+}
+
+/// Returns the URL to where the file will be served.
+
+QUrl QXmppHttpUploadSlotIq::getUrl() const
+{
+    return d->getUrl;
+}
+
+/// Sets the download URL.
+
+void QXmppHttpUploadSlotIq::setGetUrl(const QUrl &getUrl)
+{
+    d->getUrl = getUrl;
+}
+
+/// Returns a map of header fields (header name -> content) that need to be
+/// included in the PUT (upload) request. This won't contain any other fields
+/// than: "Authorization", "Cookie" or "Expires".
+
+QMap<QString, QString> QXmppHttpUploadSlotIq::headerFields() const
+{
+    return d->headerFields;
+}
+
+/// Sets the header fields the client needs to include in the PUT (upload)
+/// request. All fields other than "Authorization", "Cookie" or "Expires" will
+/// be ignored.
+
+void QXmppHttpUploadSlotIq::setHeaderFields(const QMap<QString, QString> &headerFields)
+{
+    d->headerFields.clear();
+    for (QString &name : headerFields.keys()) {
+        if (name == "Authorization" || name == "Cookie" || name == "Expires") {
+            QString value = headerFields[name];
+            d->headerFields[name] = value.replace("\n", "");
+        }
+    }
+}
+
+/// Returns whether the URLs (put and get) are both HTTPS URLs. It will return
+/// false, if the URLs are non-HTTPS URLs or not set. HTTP URLs are not
+/// XEP-compliant, but are used in some cases.
+
+bool QXmppHttpUploadSlotIq::hasHttpsUrls() const
+{
+    return d->getUrl.toString().startsWith("https://") &&
+           d->putUrl.toString().startsWith("https://");
+}
+
+bool QXmppHttpUploadSlotIq::isHttpUploadSlotIq(const QDomElement &element)
+{
+    if (element.tagName() == "iq") {
+        QDomElement slot = element.firstChildElement("slot");
+        return !slot.isNull() && slot.namespaceURI() == ns_http_upload;
+    }
+    return false;
+}
+
+/// \cond
+void QXmppHttpUploadSlotIq::parseElementFromChild(const QDomElement &element)
+{
+    QDomElement slot = element.firstChildElement("slot");
+    QDomElement put = slot.firstChildElement("put");
+    d->getUrl = QUrl::fromEncoded(slot.firstChildElement("get").attribute("url").toUtf8());
```

Comment:
This is slightly confusing, why are we not handling get / put in a symetric fashion? Also is either of these URLs optional?